Overview-of-the-Oppia-codebase.md

```diff
@@ -1,57 +1,322 @@
+
+```python
+# Handler accepts arguments from the HTTP request
+def handler(self, argument_1, argument_2):
+    # Call some services
+    value_1 = service_1(argument_1)
+    value_2 = service_2(argument_2)
+
+    # Do a little logic to combine the results from the services
+    value = (value_1 + value_2) ** 2
+
+    # Return the result
+    return self.render_json({"value": value})
+```
+
+Controllers are stored in `core/controllers/`. `main.py` defines which controller should handle each request.
+
+### Service layer
+
+Services get called by controller functions or other services to perform various computations. They often need to access Oppia's stored data, which they do by calling functions in the domain and storage layers.  We'll look at those next.
+
+Services are, confusingly, defined by files in the `core/domain/` folder. Services end in `_services.py`.
+
+### Models
+
+Before we talk about the domain layer, we need to discuss models.
+
+A model is a data structure, often a class, that stores the information that defines a particular object in Oppia. For example, we have a model for each exploration. (Note that we use "model" to refer to both the class that defines the exploration model and each instance of that class. Sorry, we know it's confusing, but this language is all over the code base.)
+
+There are many different kinds of models in Oppia. We have frontend models, backend (domain) models, and storage models.
+
+### Domain layer
+
+The domain layer defines the domain, or backend, models. These models are very similar to the frontend models, and they are what most of the backend code operates on. Each model is defined as a class.
+
+These classes don't just store information. They also provide `from_dict` and `to_dict` methods to convert between classes and dictionaries (which can be converted to JSON), and they can contain validation and converters to convert between different versions of the model.
+
+The domain models are defined in `*_domain.py` files under `core/domain/`.
+
+### Storage layer
+
+Finally, we have the storage layer, which defines the storage models.  These are also classes, but they define how data is stored in whatever system we are using to store data to the file system. In production, we use the [Google Cloud Datastore](https://cloud.google.com/datastore), and we interface with it from Python using [Cloud NDB](https://googleapis.dev/python/python-ndb/latest/index.html). (Cloud NDB is for Python 3. For Python 2, we use App Engine NDB.)
+
+Ideally we'd like all services to access the datastore through the domain layer, but some services access the storage layer directly. While you shouldn't write any new code like this, you might run into old code that works this way.
```

Comment:
This is a bit confusing, it is completely normal to interact with the storage through services, we just want to generate the domain object from the datastore class as soon as possible.

Reply:
Thanks for clarifying. I've deleted this